File: Blind 75/11 - Graphs/417-pacific-atlantic-water-flow.py
'''
There is an m x n rectangular island that borders both the Pacific Ocean and Atlantic Ocean. The Pacific Ocean touches the island's left and top edges, and the Atlantic Ocean touches the island's right and bottom edges.

The island is partitioned into a grid of square cells. You are given an m x n integer matrix heights where heights[r][c] represents the height above sea level of the cell at coordinate (r, c).

The island receives a lot of rain, and the rain water can flow to neighboring cells directly north, south, east, and west if the neighboring cell's height is less than or equal to the current cell's height. Water can flow from any cell adjacent to an ocean into the ocean.

Return a 2D list of grid coordinates result where result[i] = [ri, ci] denotes that rain water can flow from cell (ri, ci) to both the Pacific and Atlantic oceans.
'''
from time import time
from typing import List


class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(heights), len(heights[0])
        pac_seen, atl_seen = set(), set()

        def dfs(r, c, prev_h, seen):
            if (
                r < 0
                or c < 0
                or r >= ROWS
                or c >= COLS
                or (r, c) in seen
                or heights[r][c] < prev_h
            ):
                return

            seen.add((r, c))

            dfs(r + 1, c, heights[r][c], seen)
            dfs(r - 1, c, heights[r][c], seen)
            dfs(r, c + 1, heights[r][c], seen)
            dfs(r, c - 1, heights[r][c], seen)

            return

        for r in range(ROWS):
            dfs(r, 0, heights[r][0], pac_seen)
            dfs(r, COLS - 1, heights[r][COLS - 1], atl_seen)

        for c in range(COLS):
            dfs(0, c, heights[0][c], pac_seen)
            dfs(ROWS - 1, c, heights[ROWS - 1][c], atl_seen)

        return list(pac_seen & atl_seen)

    def reference(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(heights), len(heights[0])
        pacific, atlantic = set(), set()

        def dfs(r, c, visited, previous_height):
            is_in_bounds = r in range(ROWS) and c in range(COLS)
            if (
                (r, c) in visited
                or not is_in_bounds
                or heights[r][c] < previous_height
            ):
                return

            visited.add((r, c))
            directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            for dr, dc in directions:
                dfs(r + dr, c + dc, visited, heights[r][c])

        for c in range(COLS):
            dfs(0, c, pacific, heights[0][c])
            dfs(ROWS - 1, c, atlantic, heights[ROWS - 1][c])

        for r in range(ROWS):
            dfs(r, 0, pacific, heights[r][0])
            dfs(r, COLS - 1, atlantic, heights[r][COLS - 1])

        res = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) in pacific and (r, c) in atlantic:
                    res.append((r, c))

        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.pacificAtlantic(case))
                else:
                    self.pacificAtlantic(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [
            [1, 2, 2, 3, 5],
            [3, 2, 3, 4, 4],
            [2, 4, 5, 3, 1],
            [6, 7, 1, 4, 5],
            [5, 1, 1, 2, 4],
        ],
        [[2, 1], [1, 2]],
        [[1]],
    ]
    test.quantify(test_cases)

File: Blind 75/11 - Graphs/JavaScript/417-pacific-atlantic-water-flow.js
const { performance } = require('perf_hooks');

class Solution {
  method(heights) {
    if (!heights) {
      return [];
    }

    const [ROWS, COLS] = [heights.length, heights[0].length];
    let [pacific, atlantic] = [new Set(), new Set()];

    const dfs = (r, c, seen, prevHeight) => {
      if (
        seen.has(`${r}, ${c}`) ||
        r < 0 ||
        c < 0 ||
        r >= ROWS ||
        c >= COLS ||
        heights[r][c] < prevHeight
      )
        return;

      seen.add(`${r}, ${c}`);
      dfs(r + 1, c, seen, heights[r][c]);
      dfs(r - 1, c, seen, heights[r][c]);
      dfs(r, c + 1, seen, heights[r][c]);
      dfs(r, c - 1, seen, heights[r][c]);
    };

    for (let i = 0; i < ROWS; i++) {
      dfs(i, 0, pacific, heights[i][0]);
      dfs(i, COLS - 1, atlantic, heights[i][COLS - 1]);
    }

    for (let j = 0; j < COLS; j++) {
      dfs(0, j, pacific, heights[0][j]);
      dfs(ROWS - 1, j, atlantic, heights[ROWS - 1][j]);
    }

    let res = [];
    for (let i = 0; i < ROWS; i++) {
      for (let j = 0; j < COLS; j++) {
        if (pacific.has(`${i}, ${j}`) && atlantic.has(`${i}, ${j}`))
          res.push([i, j]);
      }
    }

    return res;
  }

  reference(heights) {
    const [pacificReachable, atlanticReachable] =
      this.search(heights); /* Time O(ROWS * COLS) | Space O(ROWS * COLS) */

    return this.searchGrid(
      heights,
      pacificReachable,
      atlanticReachable
    ); /* Time O(ROWS * COLS) | Space O(ROWS * COLS) */
  }

  search = (heights) => {
    const [rows, cols] = [heights.length, heights[0].length];
    const [pacificReachable, atlanticReachable] = [
      this.getMatrix(rows, cols),
      this.getMatrix(rows, cols),
    ]; /* Time O(ROWS * COLS) | Space O(ROWS * COLS) */

    this.searchRows(heights, rows, cols, pacificReachable, atlanticReachable);
    this.searchCols(heights, rows, cols, pacificReachable, atlanticReachable);

    return [pacificReachable, atlanticReachable];
  };

  searchGrid = (
    heights,
    pacificReachable,
    atlanticReachable,
    intersection = []
  ) => {
    const [rows, cols] = [heights.length, heights[0].length];

    for (let row = 0; row < rows; row++) {
      /* Time O(ROWS) */
      for (let col = 0; col < cols; col++) {
        /* Time O(COLS) */
        const isReachable =
          pacificReachable[row][col] && atlanticReachable[row][col];
        if (!isReachable) continue;

        intersection.push([row, col]); /* Space O(ROWS * COLS) */
      }
    }

    return intersection;
  };

  getMatrix = (rows, cols) =>
    new Array(rows)
      .fill() /* Time O(ROWS * COLS) | Space O(ROWS * COLS) */
      .map(() => new Array(cols).fill(false));

  searchRows = (heights, rows, cols, pacificReachable, atlanticReachable) => {
    for (let row = 0; row < rows; row++) {
      /* Time O(ROWS) */
      const [pacificStart, atlanticStart] = [0, cols - 1];

      this.dfs(
        row,
        pacificStart,
        rows,
        cols,
        pacificReachable,
        heights
      ); /* Space O(ROWS * COLS) */
      this.dfs(
        row,
        atlanticStart,
        rows,
        cols,
        atlanticReachable,
        heights
      ); /* Space O(ROWS * COLS) */
    }
  };

  searchCols = (heights, rows, cols, pacificReachable, atlanticReachable) => {
    for (let col = 0; col < cols; col++) {
      /* Time O(COLS) */
      const [pacificStart, atlanticStart] = [0, rows - 1];

      this.dfs(
        pacificStart,
        col,
        rows,
        cols,
        pacificReachable,
        heights
      ); /* Space O(ROWS * COLS) */
      this.dfs(
        atlanticStart,
        col,
        rows,
        cols,
        atlanticReachable,
        heights
      ); /* Space O(ROWS * COLS) */
    }
  };

  dfs = (row, col, rows, cols, isReachable, heights) => {
    isReachable[row][col] = true;

    for (const [_row, _col] of this.getNeighbors(row, rows, col, cols)) {
      if (isReachable[_row][_col]) continue;

      const isLower = heights[_row][_col] < heights[row][col];
      if (isLower) continue;

      this.dfs(
        _row,
        _col,
        rows,
        cols,
        isReachable,
        heights
      ); /* Space O(ROWS * COLS) */
    }
  };

  getNeighbors = (row, rows, col, cols) =>
    [
      [0, 1],
      [0, -1],
      [1, 0],
      [-1, 0],
    ]
      .map(([_row, _col]) => [row + _row, col + _col])
      .filter(
        ([_row, _col]) => 0 <= _row && _row < rows && 0 <= _col && _col < cols
      );

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [1, 2, 2, 3, 5],
    [3, 2, 3, 4, 4],
    [2, 4, 5, 3, 1],
    [6, 7, 1, 4, 5],
    [5, 1, 1, 2, 4],
  ],
  [
    [2, 1],
    [1, 2],
  ],
];
test.quantify(testCases);
